File: problems/javascript/280721-intersection-of-two-linked-list/code.js
/**
 * https://leetcode.com/problems/intersection-of-two-linked-lists/
 *
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
const getIntersectionNode = (headA, headB) => {
  const visitedNodeA = []

  let traveller = headA
  while (traveller) {
    visitedNodeA.push(traveller)
    traveller = traveller.next
  }

  traveller = headB
  while (traveller) {
    if (visitedNodeA.includes(traveller)) return traveller
    traveller = traveller.next
  }

  return null
}

File: problems/ruby/28072021-intersection-of-two-linked-lists/code.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# https://leetcode.com/problems/intersection-of-two-linked-lists/
#
# @param {ListNode} headA
# @param {ListNode} headB
# @return {ListNode}
def getIntersectionNode(headA, headB)
  visited_node = {}

  traveller = headA
  while traveller
    visited_node[traveller] = true
    traveller = traveller.next
  end

  traveller = headB
  while traveller
    return traveller if visited_node[traveller]
    traveller = traveller.next
  end

  nil
end
